File: python/day_03.py
#!/usr/bin/env python3

# inspired by the version by https://github.com/Eixix/Advent-of-Code
def main():
    f_name = "data/input.txt"

    infile = open(f_name)

    n = 0

    # assuming always 12 cols
    ncol = 12
    count_list = [0] * ncol

    # assuming always 1000 rows
    nrow = 1000

    for line in infile:
        for i in range(ncol):
            count_list[i] += int(line[i])

    gamma_bits = [int(x > (nrow / 2)) for x in count_list]

    epsilon_bits = [-(x - 1) for x in gamma_bits]

    gamma   = int("".join(map(str, gamma_bits)),   2)
    epsilon = int("".join(map(str, epsilon_bits)), 2)

    print(f"Power consumption is: {gamma * epsilon}")

if __name__ == "__main__":
    main()

File: R/day_03.R
library("magrittr")

# convert input to numeric matrix
input_mat <- readLines("../data/input.txt") %>% 
  strsplit("") %>% 
  lapply(as.numeric) %>% 
  unlist() %>% 
  matrix(nrow = 1000, ncol = 12, byrow = TRUE)

# extract mode of each column for the gamma frequency bits
gamma_bits <- apply(input_mat, 2, function(col) {
  tab <- table(col)
  tab[tab == max(tab)] %>% 
    names() %>% 
    as.numeric()
})

# invert gamma frequency bits
# wouldn't have noticed you can just do this instead of applying the above 
#   function using min if it wasn't for you, Tobi
epsilon_bits <- -(gamma_bits - 1)

# convert bits to number
# nicked from https://stackoverflow.com/a/69819329
gamma   <- sum(gamma_bits   * 2 ^ ((length(gamma_bits  ) - 1):0))
epsilon <- sum(epsilon_bits * 2 ^ ((length(epsilon_bits) - 1):0))

print(paste("Power consumption is:", gamma * epsilon))